keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1003 @@
+the local state, change frequency is reduced compared to cluster state. Testing on actual clusters
+showed a frequency reduction of change events by 2 orders of magnitude.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+TBD (Calico eBPF)
+
+#### Story 2
+
+TBD (node-local cluster DNS provider)
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- sending the full-state could be resource consuming on big clusters, but it should still be O(1) to
+  the actual kernel definitions (the complexity of what the node has to handle cannot be reduced
+  without losing functionality or correctness).
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A [draft implementation] exists and some [performance testing] has been done.
+
+[draft implementation]: https://github.com/kubernetes-sigs/kpng/
+[performance testing]: https://github.com/mcluseau/kube-proxy2/blob/master/doc/proposal.md
+
+The watchable API will be a long polling, taking a "last known state info" and returning a stream of
+objects. 
+
+Proposed definition is found here: https://github.com/kubernetes-sigs/kpng/blob/master/api/localnetv1/services.proto
+
+The main types are:
+```
+message Service
+message IPFilter
+message ServiceIPs
+message Endpoint
+message IPSet
+message Port
+message ClientIPAffinity
+message ServiceInfo
+message EndpointInfo
+message EndpointConditions
+message NodeInfo
+message Node
+```
+
+When the proxy starts, it will generate a random InstanceID, and have Rev at 0. So, a client
+(re)connecting will get the new state either after a proxy restart or when an actual change occurs.
+The proxy will never send a partial state, only full states. This means it waits to have all its
+Kubernetes watchers sync'ed before going to Rev 1.
+
+The first NextItem in the stream will be the state info required for the next polling call, and any
```

Comment:
This seems to be out of date since neither the above text nor the linked API has anything called "NextItem"?

Reply:
yes it's `OpItem` now (https://github.com/kubernetes-sigs/kpng/blob/master/api/localnetv1/services.proto#L23)

Reply:
Yeah, the detalis are : 

Backends add themselves to LocalCmd
LocalCmd gets the “Sink” from the backend (each backend implements Sink() function)
The Sink in the backend implements
Setup, WaitRequest, Reset, Send
Send types are “Sync, Reset, Set, Delete”
client/localsink/decoder/decoder.go has the logic converting OpItems into “SetService” . 

ill add this

Reply:
i added a TODO and cleaned some of this